src/core/linalg/src/dense/4C_linalg_utils_densematrix_funct.hpp

```diff
@@ -0,0 +1,307 @@
+// This file is part of 4C multiphysics licensed under the
+// GNU Lesser General Public License v3.0 or later.
+//
+// See the LICENSE.md file in the top-level for license information.
+//
+// SPDX-License-Identifier: LGPL-3.0-or-later
+
+/*----------------------------------------------------------------------*/
+/*! \file
+
+\brief A collection of methods to evaluate matrix functions, such as the square root, the
+exponential and the logarithm, along with specific derivatives in namespace Core::LinAlg
+
+\level 0
+*/
+/*----------------------------------------------------------------------*/
+
+#ifndef FOUR_C_LINALG_UTILS_DENSEMATRIX_FUNCT_HPP
+#define FOUR_C_LINALG_UTILS_DENSEMATRIX_FUNCT_HPP
+
+#include "4C_config.hpp"
+
+#include "4C_linalg_fixedsizematrix.hpp"
+#include "4C_linalg_utils_densematrix_eigen.hpp"
+
+#include <magic_enum/magic_enum.hpp>
+
+#include <cstddef>
+#include <optional>
+
+FOUR_C_NAMESPACE_OPEN
+
+namespace Core::LinAlg
+{
+  /// enum class: error types for the calculation of matrix functions
+  enum class MatrixFunctErrorType
+  {
+    no_errors,           // evaluation without errors
+    unsuitable_method,   // unsuitable computation method (e.g., the matrix norm is too large to use
+                         // a Taylor series description)
+    failed_computation,  // although the computation method is suitable, the evaluation fails
+                         // nonetheless (e.g., the maximum number of
+                         // iterations is exceeded)
+  };
+
+  /// make sure MatrixFunctErrorType is stream-insertable
+  inline std::ostream& operator<<(
+      std::ostream& stream, const MatrixFunctErrorType& matrix_funct_err_type)
+  {
+    stream << magic_enum::enum_name(matrix_funct_err_type);
+    return stream;
+  }
```

Comment:
Let's do this differently: I will soon add a header `4C_utils_enum.hpp` which provides such convenience magically without requiring extra definitions. You would only need to include the header in the file where you want to print. The `magic_enum` header is rather heavy so we don't want to pull it in if we simply need the enums here.